lib/vtls/openssl.c

```diff
@@ -2505,6 +2505,61 @@ static int ossl_new_session_cb(SSL *ssl, SSL_SESSION *ssl_sessionid)
   return res;
 }
 
+static CURLcode load_cacert_from_memory(SSL_CTX *ctx,
+                                        const char * const ca_info_pem)
+{
+  /* these need freed at the end */
+  BIO *cbio = NULL;
+  STACK_OF(X509_INFO) *inf = NULL;
+
+  /* everything else is just a reference */
+  int i, count = 0;
+  X509_STORE *cts = NULL;
+  X509_INFO *itmp = NULL;
+
+  CURLcode result = CURLE_SSL_CACERT_BADFILE;
+
+  do {
+    /* casting strlen to int feels wrong, but what to do with this API? */
+    cbio = BIO_new_mem_buf(ca_info_pem, (int) strlen(ca_info_pem));
+    if(!cbio)
+      return CURLE_OUT_OF_MEMORY;
+
+    cts = SSL_CTX_get_cert_store(ctx);
+    if(!cts)
+      break;
+
+    inf = PEM_X509_INFO_read_bio(cbio, NULL, NULL, NULL);
+    if(!inf)
+      break;
+
+    /* add each entry from PEM file to x509_store */
+    for(i = 0; i < (int)sk_X509_INFO_num(inf); ++i) {
+      itmp = sk_X509_INFO_value(inf, i);
+      if(itmp->x509) {
+        X509_STORE_add_cert(cts, itmp->x509);
+        ++count;
+      }
+      if(itmp->crl) {
+        X509_STORE_add_crl(cts, itmp->crl);
+        ++count;
+      }
+    }
+
+    /* if we didn't end up importing anything, treat that as an error */
+    if(count > 0)
+      result = CURLE_OK;
```

Comment:
I find this comment slightly misleading since the return code of the import functions is not checked. I don't think it should fail if there was nothing to import in the certificate, and I wonder if it should fail at all since other certificates may be read in. Maybe make it always a warning instead of just in the insecure case. Interested in what others think about this.

Reply:
Here's another thing, every time an [X509_store_add_](https://www.openssl.org/docs/man1.1.1/man3/X509_STORE_add_crl.html) is called it increases the reference count, so what if it's called twice don't you have to free two objects x509 and crl?

_"X509_STORE_add_cert() and X509_STORE_add_crl() add the respective object to the X509_STORE's local storage. Untrusted objects should not be added in this way. The added object's reference count is incremented by one, hence the caller retains ownership of the object and needs to free it when it is no longer needed."_

Reply:
> 
> I find this comment slightly misleading since the return code of the import functions is not checked.
This was true on first version, but now load_cacert_from_memory value is checked (as requester by @bagder ).

But I think we must remove this test and don't fail if there is nothing to import.

Reply:
ok, I removed the test